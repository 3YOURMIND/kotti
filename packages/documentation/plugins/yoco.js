import '@3yourmind/yoco/style.css'
