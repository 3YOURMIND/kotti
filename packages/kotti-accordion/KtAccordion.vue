<template>
	<div class="accordion">
		<div class="accordion__header">
			<div class="accordion__title">
				<slot name="title">
					<i v-if="icon" v-text="icon" class="yoco accordion__title__icon" />
					<div class="accordion__title__text">{{ title }}</div>
				</slot>
			</div>
			<div @click="toggle" class="accordion__toggle">
				<i v-text="isOpen ? 'minus' : 'plus'" class="yoco" />
			</div>
		</div>
		<div
			:style="`--height: ${contentHeight}px`"
			:class="isOpen ? 'is-open' : 'is-close'"
			@click="setHeight"
			class="accordion__content"
		>
			<div ref="contentInner" class="inner">
				<slot />
			</div>
		</div>
	</div>
</template>

<script>
export default {
	name: 'KtAccordion',
	props: {
		title: {
			type: String,
			required: true,
		},
		icon: {
			type: String,
		},
		isClosed: {
			type: Boolean,
			default: false,
		},
	},
	data() {
		return {
			isOpen: true,
			contentHeight: null,
			interval: null,
		}
	},
	mounted() {
		this.isOpen = !this.isClosed
		// This is the best solution right now
		// It will handle the case when the content change
		const reloadingTime = 500
		this.interval = setInterval(this.setHeight, reloadingTime)
	},
	beforeDestroy() {
		clearInterval(this.interval)
	},
	methods: {
		setHeight() {
			const height = this.$refs.contentInner.clientHeight
			this.contentHeight = height
		},
		toggle() {
			this.setHeight()
			this.isOpen = !this.isOpen
		},
	},
}
</script>
<style lang="scss">
.accordion {
	margin-bottom: $unit-4;
	border-bottom: none;
	border-radius: 2px;
}

.accordion__header {
	display: flex;
	justify-content: space-between;
	padding: $unit-2 $unit-8;
	border: 1px solid $lightgray-400;
}

.accordion__title {
	flex-grow: 1;
	align-self: center;
	font-size: 16px;
	font-weight: 600;
}

.accordion__content {
	overflow: hidden;
	border: 1px solid $lightgray-400;
	border-top: none;
	transition: height 200ms linear;
	.inner {
		padding: $unit-2 $unit-8;
	}
	&.is-open {
		height: var(--height);
	}
	&.is-close {
		height: 0px;
	}
}

.accordion__title__icon,
.accordion__title__text {
	display: inline-block;
}

.accordion__title__icon {
	margin-right: $unit-4;
	font-size: 22px !important;
	color: $primary-500;
}

.accordion__toggle {
	align-self: center;
	margin-left: $unit-4;
	color: $primary-500;
	cursor: pointer;
	user-select: none;
}
</style>
