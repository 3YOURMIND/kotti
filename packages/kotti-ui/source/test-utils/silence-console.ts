export const silenceConsole = (): void => {
	console.debug = () => undefined
	console.error = () => undefined
	console.group = () => undefined
	console.groupEnd = () => undefined
	console.info = () => undefined
	console.log = () => undefined
	console.warn = () => undefined
}
