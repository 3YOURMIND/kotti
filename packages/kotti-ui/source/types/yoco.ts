export namespace Yoco {
	/**
	 * Should be kept up-to-date with yoco
	 * @see {@link ./data.json}
	 */
	export type Icon =
		| '3dbox'
		| 'announce'
		| 'arrow_line'
		| 'attachment'
		| 'basket'
		| 'bell'
		| 'burger'
		| 'calendar'
		| 'card'
		| 'charts'
		| 'check'
		| 'chevron_down'
		| 'chevron_left'
		| 'chevron_right'
		| 'chevron_up'
		| 'circle_attention'
		| 'circle_check'
		| 'circle_cross'
		| 'circle_i'
		| 'circle_load'
		| 'circle_question'
		| 'clock'
		| 'close'
		| 'cloud_download'
		| 'cloud_upload'
		| 'cloud'
		| 'comment'
		| 'copy'
		| 'dashboard'
		| 'dimension'
		| 'dots_vertical'
		| 'dots'
		| 'download'
		| 'edit'
		| 'email'
		| 'exit_fullscreen'
		| 'eye'
		| 'file'
		| 'filter'
		| 'folder'
		| 'fullscreen'
		| 'global'
		| 'heart'
		| 'invoice'
		| 'layer'
		| 'list'
		| 'location'
		| 'logout'
		| 'markup'
		| 'material'
		| 'minus'
		| 'not_allow'
		| 'notes'
		| 'permission'
		| 'picture'
		| 'plus'
		| 'print'
		| 'save'
		| 'scale'
		| 'search'
		| 'setting'
		| 'share'
		| 'shipping'
		| 'shop'
		| 'sidebar'
		| 'star'
		| 'structure'
		| 'support'
		| 'tag'
		| 'tech'
		| 'trash'
		| 'triangle_down'
		| 'triangle_left'
		| 'triangle_right'
		| 'triangle_up'
		| 'upload'
		| 'user'
		| 'usergroup'
		| 'version'
		| 'view'
		| 'zip'
}
