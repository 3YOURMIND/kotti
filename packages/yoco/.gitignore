node_modules/
yarn-error.log
.DS_Store
dist/
gh-pages/