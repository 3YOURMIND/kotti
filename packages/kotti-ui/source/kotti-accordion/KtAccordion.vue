<template>
	<div class="accordion">
		<div
			class="accordion__header"
			:class="{ 'accordion__header--clickable': isFullyClickable }"
			@click="headerToggle"
		>
			<div class="accordion__title">
				<slot name="title">
					<i v-if="icon" class="yoco accordion__title__icon" v-text="icon" />
					<div class="accordion__title__text" v-text="title" />
				</slot>
			</div>
			<div class="accordion__toggle" @click.stop="toggle">
				<i class="yoco" v-text="isContentOpen ? 'minus' : 'plus'" />
			</div>
		</div>
		<div
			class="accordion__content"
			:class="isContentOpen ? 'is-open' : 'is-close'"
		>
			<div ref="contentInner" class="inner">
				<slot />
			</div>
		</div>
	</div>
</template>

<script lang="ts">
import { defineComponent, ref } from '@vue/composition-api'

import { useSlideAnimaion } from './hooks'
import { KottiAccordion } from './types'

export default defineComponent<KottiAccordion.PropsInternal>({
	name: 'KtAccordion',
	props: {
		icon: { default: null, type: String },
		isClosed: { default: false, type: Boolean },
		isFullyClickable: { default: false, type: Boolean },
		title: { required: true, type: String },
	},
	setup(props) {
		const contentInner = ref<HTMLElement | null>(null)

		const { isContentOpen, toggle } = useSlideAnimaion(contentInner, {
			duration: 300,
			isInitiallyClosed: props.isClosed,
		})
		return {
			contentInner,
			headerToggle: () => {
				if (props.isFullyClickable) toggle()
			},
			isContentOpen,
			toggle,
		}
	},
})
</script>

<style lang="scss">
:root {
	--accordion-color: var(--interactive-03);
}

@import '../kotti-style/_variables.scss';

.accordion {
	margin-bottom: var(--unit-4);
	border-radius: 2px;
}

.accordion__header {
	display: flex;
	justify-content: space-between;
	padding: var(--unit-2) var(--unit-8);
	border: 1px solid var(--ui-02);

	.yoco {
		font-size: 1.2rem;
	}
}
.accordion__header--clickable {
	cursor: pointer;
}

.accordion__title {
	display: flex;
	flex-grow: 1;
	align-items: center;
	align-self: center;
	font-size: 16px;
	font-weight: 600;
}

.accordion__content {
	margin-top: -1px; // prevent border overlap when closed
	overflow: hidden;
	border: 1px solid var(--ui-02);
	border-top: none;
	transition: height 200ms linear;

	.inner {
		padding: var(--unit-2) var(--unit-8);
	}
	&.is-open {
		height: var(--height);
	}
	&.is-close {
		height: 0px;
	}
}

.accordion__title__icon,
.accordion__title__text {
	display: inline-block;
}

.accordion__title__icon {
	margin-right: var(--unit-4);
	color: var(--accordion-color);
}

.accordion__toggle {
	display: flex;
	align-items: center;
	align-self: center;
	margin-left: var(--unit-4);
	color: var(--accordion-color);
	cursor: pointer;
	user-select: none;
}
</style>
