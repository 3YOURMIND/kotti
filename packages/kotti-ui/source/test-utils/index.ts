import VueCompositionApi from '@vue/composition-api'
import { createLocalVue } from '@vue/test-utils'

import { KottiField } from '../kotti-field/types'
import { KottiForm } from '../kotti-form/types'

/**
 * Access lazily-evaluated computed property to force it to evaluate
 * @see {@link https://github.com/vuejs/vue/issues/7672}
 */
export const forceVueToEvaluateComputedProperty = <T>(kottiFieldProp: T) =>
	kottiFieldProp

export const localVue = createLocalVue()
localVue.use(VueCompositionApi)

export const getMockContext = (
	{ validators = {}, values = {} } = { validators: {}, values: {} },
): Pick<
	KottiForm.Context,
	'fieldInheritableProps' | 'validators' | 'values'
> & {
	onAddField: jest.Mock
	onRemoveField: jest.Mock
	setValue: jest.Mock
} => ({
	// @ts-expect-error this is an imperfect mock and not a real ref
	fieldInheritableProps: {
		value: {
			hideClear: false,
			hideValidation: false,
			isDisabled: false,
			isLoading: false,
			size: KottiField.Size.MEDIUM,
		},
	},
	onAddField: jest.fn(),
	onRemoveField: jest.fn(),
	setValue: jest.fn(),
	// @ts-expect-error this is an imperfect mock and not a real ref
	validators: { value: validators },
	// @ts-expect-error this is an imperfect mock and not a real ref
	values: { value: values },
})
