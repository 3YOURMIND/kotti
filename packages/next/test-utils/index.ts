import VueCompositionApi from '@vue/composition-api'
import { createLocalVue } from '@vue/test-utils'

import { KottiForm } from '../kotti-form/types'

/**
 * Access lazily-evaluated computed property to force it to evaluate
 * @see {@link https://github.com/vuejs/vue/issues/7672}
 */
export const forceVueToEvaluateComputedProperty = <T>(kottiFieldProp: T) =>
	kottiFieldProp

export const localVue = createLocalVue()
localVue.use(VueCompositionApi)

export const getMockContext = (
	{ validators = {}, values = {} } = { validators: {}, values: {} },
): Pick<
	KottiForm.Context,
	'hideValidation' | 'isLoading' | 'validators' | 'values'
> & {
	onAddField: jest.Mock
	onRemoveField: jest.Mock
	setValue: jest.Mock
} => ({
	hideValidation: { value: false },
	isLoading: { value: false },
	onAddField: jest.fn(),
	onRemoveField: jest.fn(),
	setValue: jest.fn(),
	validators: { value: validators },
	values: { value: values },
})
