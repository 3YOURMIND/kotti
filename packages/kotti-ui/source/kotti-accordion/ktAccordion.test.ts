import { shallowMount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'

import { Yoco } from '@3yourmind/yoco'

import KtAccordion from './KtAccordion.vue'

describe('KtAccordion.vue', () => {
	it('renders correctly', () => {
		// @ts-expect-error mount type is not compatible with return type from defineComponent
		const wrapper = shallowMount(KtAccordion)

		expect(wrapper.element).toMatchSnapshot()
	})

	it('renders correctly (with props)', () => {
		// @ts-expect-error mount type is not compatible with return type from defineComponent
		const wrapper = shallowMount(KtAccordion, {
			propsData: {
				dataTest: 'my-accordion',
				icon: Yoco.Icon.BOOKMARK,
				isClosed: true,
				title: 'My Accordion',
			},
		})

		expect(wrapper.element).toMatchSnapshot()
	})
})
