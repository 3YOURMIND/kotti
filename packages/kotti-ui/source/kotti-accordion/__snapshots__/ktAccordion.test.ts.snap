// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`KtAccordion.vue > renders correctly (with props) 1`] = `
<div
  class="kt-accordion"
  data-v-b260b1b5=""
>
  <div
    class="kt-accordion__header"
    data-test="my-accordion"
    data-v-b260b1b5=""
  >
    <div
      class="kt-accordion__title"
      data-v-b260b1b5=""
    >
      <i
        class="yoco kt-accordion__icon"
        data-v-b260b1b5=""
      >
        bookmark
      </i>
      <div
        class="kt-accordion__text"
        data-v-b260b1b5=""
      >
        My Accordion
      </div>
    </div>
    <div
      class="kt-accordion__toggle"
      data-v-b260b1b5=""
    >
      <i
        class="yoco"
        data-v-b260b1b5=""
      >
        plus
      </i>
    </div>
  </div>
  <div
    class="kt-accordion__content kt-accordion__content--is-closed"
    data-v-b260b1b5=""
  >
    <div
      class="kt-accordion__inner"
      data-v-b260b1b5=""
    />
  </div>
</div>
`;

exports[`KtAccordion.vue > renders correctly 1`] = `
<div
  class="kt-accordion"
  data-v-b260b1b5=""
>
  <div
    class="kt-accordion__header"
    data-v-b260b1b5=""
  >
    <div
      class="kt-accordion__title"
      data-v-b260b1b5=""
    >
      <!---->
      <div
        class="kt-accordion__text"
        data-v-b260b1b5=""
      />
    </div>
    <div
      class="kt-accordion__toggle"
      data-v-b260b1b5=""
    >
      <i
        class="yoco"
        data-v-b260b1b5=""
      >
        minus
      </i>
    </div>
  </div>
  <div
    class="kt-accordion__content kt-accordion__content--is-open"
    data-v-b260b1b5=""
  >
    <div
      class="kt-accordion__inner"
      data-v-b260b1b5=""
    />
  </div>
</div>
`;
